src/inference/src/dev/core_impl.cpp

```diff
@@ -1242,26 +1280,28 @@ ov::AnyMap ov::CoreImpl::create_compile_config(const ov::Plugin& plugin, const o
     }
 
     // 1. Move DEVICE_ID key to property_config
+    const bool supports_device_id = device_supports_property(plugin, ov::device::id);
     auto deviceIt = user_config.find(ov::device::id.name());
     if (deviceIt != user_config.end()) {
         property_config[deviceIt->first] = deviceIt->second.as<std::string>();
-    } else {
-        // we likely need to extract default device_id from the plugin,
-        // but we suppose when we call plugin.get_property it will provide the answer
-        // for the default device (e.g. DEVICE_ID = 0 for GPU)
+    } else if (supports_device_id) {
+        property_config[ov::device::id.name()] = plugin.get_property(ov::device::id, {});
     }
 
     // 2. Replace DEVICE_ID with DEVICE_ARCHITECTURE value to identify device
-    if (device_supports_property(plugin, ov::device::architecture.name())) {
+    if (device_supports_property(plugin, ov::device::architecture)) {
         compile_config[ov::device::architecture.name()] =
             plugin.get_property(ov::device::architecture, property_config);
     } else {
-        // Take device name if device does not support DEVICE_ARCHITECTURE metric
-        compile_config[ov::device::architecture.name()] = plugin.get_name();
+        // Take full device name if device does not support DEVICE_ARCHITECTURE metric
+        std::string full_device_name = plugin.get_name();
+        if (supports_device_id)
+            full_device_name += "." + property_config[ov::device::id.name()].as<std::string>();
+        compile_config[ov::device::architecture.name()] = full_device_name;
```

Comment:
full_device_name will also impact model blob cache? GPU.0 and GPU.1 will lead to different compiled blobs, right?

Reply:
`compile_config[ov::device::architecture.name()] `is only one config which may impact hash of model cache, if other all configs are same except for full_device_name, the hash will be different. in this point, full_device_name in this branch will lead to different model cache.

Reply:
>  in this point, full_device_name in this branch will lead to different model cache.

Yes, this is an intention. 
Because if you have some unknown `DEVICE` with IDs `X` and `Y` - you don't know whether they have different architecture or not. If you ignore device_id, them a single blob can be compiled and can cause issues if the devices are really not compatible. 
So, it's safer to generate 2 blobs because of 2 devices X and Y.

If `DEVICE` wants to avoid duplication of compiled blobs, it can support device::architecture (which is simple change). In this case, the current branch of code is ignored.

Are you ok with such arguments?

Reply:
It seems only gpu plugin and hetero plugin support ov::device::architecture, and hetero plugin is virtual plugin which will not use it. 

I'm not sure what's unknown DEVICE, but if we use device id to identify them, it means we always create different blob cache for different non-GPU devices, such as for VPU/GNA,  why not use ov::device::full_name? or add support device::architecture for all hw device?

Reply:
I've extracted caching related changes to a dedicated PR https://github.com/openvinotoolkit/openvino/pull/16528
and will revert changes from this one. 

Please, review that PR as well. I hope the logic is more clear now.